### ISSUE TITLE
Results do not accord with sklearn GMMs

### ISSUE THREAD

**Issue by morawi** [borchero/pycave#6]
I tried to compare the results with sklearn GMM. Unfortunately, results do not accord. Not sure if I'm missing something? 
Here's the code, data file (numpy format) is attached [3d_array.zip](https://github.com/borchero/pycave/files/5906035/3d_array.zip)

`
from pycave.bayes.gmm import GMM
import torch
import numpy as np
from sklearn.mixture import GaussianMixture

X = np.load(' 3d_array.npy')
X = X.astype('float32')
X1 = X[0:3]

n_components=2
print('Sklearn Run')
skl_model = GaussianMixture(n_components=n_components, random_state=0).fit(X)
print('means', skl_model.means_)
labels = skl_model.predict(X1)
print('Prediction labels of X1:', labels)

config = {   
            'num_components': n_components,
            'num_features': X.shape[1],
            'covariance': 'spherical', # {'spherical', 'diag-shared'}            
        }

print('\n Pycave Run')
torch.manual_seed(0)
model = GMM(**config)
model.fit(torch.tensor(X1))
print('means', model.gaussian.means.cpu().numpy())
labels = model.predict(torch.tensor(X1)).cpu().numpy()
print('Prediction labels of X1:', np.argmax(labels, axis=1))

`

**Comment by morawi** [comment]
I think the above code was not thoughtful, hence, I've done an extensive test. Again, PyCave does not accord, not only with Sklearn, but with itself. Unless I'm doing something extremely wrong. Here's the test:

```
from pycave.bayes.gmm import GMM
import torch
import numpy as np
from sklearn.mixture import GaussianMixture


n_components=2
X = np.load('C:\\MyPrograms\\3d_array.npy').astype('float32')

config = {   
            'num_components': n_components,
            'num_features': X.shape[1],
            'covariance': 'spherical', # {'spherical', 'diag-shared'}            
        }


def pycave_vs_sklearn_gmm(X, config):
    # print('Sklearn Run ... ')    
    skl_model = GaussianMixture(n_components = n_components, 
                                covariance_type = config['covariance'],
                                random_state = np.random.randint(0, 1e9),
                                ).fit(X)
    # print('means', skl_model.means_)
    labels_sk = skl_model.predict(X)
    
    # print('\n Pycave Run ... ')
    torch.manual_seed(np.random.randint(0, 1e9)) # # torch.random.seed() 
    model = GMM(**config)
    # model.reset_parameters(torch.tensor(X))
    model.fit(torch.tensor(X))
    # print('means', model.gaussian.means.cpu().numpy())
    labels = np.argmax(model.predict(torch.tensor(X)).cpu().numpy(), axis=1)

    # Check accordance
    return (sum(labels_sk==labels) or sum(labels_sk!=labels) ) / len(X) 

def pycave_vs_pycave_gmm(X, config):
    
    torch.manual_seed(np.random.randint(0, 1e9)) # # torch.random.seed() 
    model = GMM(**config)
    # model.reset_parameters(torch.tensor(X))
    model.fit(torch.tensor(X))
    # print('means', model.gaussian.means.cpu().numpy())
    labels = np.argmax(model.predict(torch.tensor(X)).cpu().numpy(), axis=1)
    
    torch.manual_seed(np.random.randint(0, 1e9)) # # torch.random.seed() 
    model2 = GMM(**config)
    # model.reset_parameters(torch.tensor(X))
    model2.fit(torch.tensor(X))
    # print('means', model.gaussian.means.cpu().numpy())
    labels2 = np.argmax(model2.predict(torch.tensor(X)).cpu().numpy(), axis=1)


    # Check accordance
    return (sum(labels2==labels) or sum(labels2!=labels) ) / len(X) 


def sklearn_vs_sklearn_gmm(X, config):
    # print('Sklearn Run ... ')    
    skl_model = GaussianMixture(n_components = n_components, 
                                covariance_type = config['covariance'],
                                random_state = np.random.randint(0, 1e9),
                                ).fit(X)
    # print('means', skl_model.means_)
    labels_sk = skl_model.predict(X)
    
    skl_model2 = GaussianMixture(n_components = n_components, 
                                covariance_type = config['covariance'],
                                random_state = np.random.randint(0, 1e9),
                                ).fit(X)
    # print('means', skl_model.means_)
    labels_sk2 = skl_model2.predict(X)
    
    # Check accordance
    return (sum(labels_sk==labels_sk2) or sum(labels_sk!=labels_sk2) ) / len(X) 



accord= []; sk_sk_accord = []; pycav_accord =[]
for i in range(10):
    accord.append(pycave_vs_sklearn_gmm(X, config))
    sk_sk_accord.append(sklearn_vs_sklearn_gmm(X, config))
    pycav_accord.append(pycave_vs_pycave_gmm(X, config))

print('Accordance over 10 experiments')
print("Sklearn vs Sklearn GMM Accordance :  % 2.2f %%" %(100*np.mean(sk_sk_accord)))
print("PyCave vs Pycave GMM Accordance :  % 2.2f %%" %(100*np.mean(accord)))
print("PyCave vs Sklearn GMM Accordance :  % 2.2f %%" %(100*np.mean(pycav_accord)))
```

```
Accordance over 10 experiments
Sklearn vs Sklearn GMM Accordance :   100.00 %
PyCave vs Pycave GMM Accordance :   40.10 %
PyCave vs Sklearn GMM Accordance :   55.12 %

```

Sklearn GMM accords with itself, PyCave GMM does not accord with itself nor sklearn.

**Comment by morawi** [comment]
I think the above test does not measure the accordance well. Hence, I did the one below. I also measured inter and intra GMM accordance, ie, sklearn vs sklearn, pycave vs pycave. Here's the code and the results:


```
from pycave.bayes.gmm import GMM
import torch
import numpy as np
from sklearn.mixture import GaussianMixture


n_components=2
X = np.load('C:\\MyPrograms\\3d_array.npy').astype('float32')

config = {   
            'num_components': n_components,
            'num_features': X.shape[1],
            'covariance': 'spherical', # {'spherical', 'diag-shared'}            
        }


def pycave_vs_sklearn_gmm(X, config):
    # print('Sklearn Run ... ')    
    skl_model = GaussianMixture(n_components = n_components, 
                                covariance_type = config['covariance'],
                                random_state = np.random.randint(0, 1e9),
                                ).fit(X)
    # print('means', skl_model.means_)
    labels_sk = skl_model.predict(X)
    
    # print('\n Pycave Run ... ')
    torch.manual_seed(np.random.randint(0, 1e9)) # # torch.random.seed() 
    model = GMM(**config)
    # model.reset_parameters(torch.tensor(X))
    model.fit(torch.tensor(X))
    # print('means', model.gaussian.means.cpu().numpy())
    labels = np.argmax(model.predict(torch.tensor(X)).cpu().numpy(), axis=1)

    # Check accordance, taking the max in case labels are swapped
    return max(sum(labels_sk==labels), sum(labels_sk!= labels) ) / len(X) 

def pycave_vs_pycave_gmm(X, config):
    
    torch.manual_seed(np.random.randint(0, 1e9)) # # torch.random.seed() 
    model = GMM(**config)
    # model.reset_parameters(torch.tensor(X))
    model.fit(torch.tensor(X))
    # print('means', model.gaussian.means.cpu().numpy())
    labels = np.argmax(model.predict(torch.tensor(X)).cpu().numpy(), axis=1)
    
    torch.manual_seed(np.random.randint(0, 1e9)) # # torch.random.seed() 
    model2 = GMM(**config)
    # model.reset_parameters(torch.tensor(X))
    model2.fit(torch.tensor(X))
    # print('means', model.gaussian.means.cpu().numpy())
    labels2 = np.argmax(model2.predict(torch.tensor(X)).cpu().numpy(), axis=1)


    # Check accordance, taking the max in case labels are swapped
    return max(sum(labels2==labels) , sum(labels2!=labels) ) / len(X) 


def sklearn_vs_sklearn_gmm(X, config):
    # print('Sklearn Run ... ')    
    skl_model = GaussianMixture(n_components = n_components, 
                                covariance_type = config['covariance'],
                                random_state = np.random.randint(0, 1e9),
                                ).fit(X)
    # print('means', skl_model.means_)
    labels_sk = skl_model.predict(X)
    
    skl_model2 = GaussianMixture(n_components = n_components, 
                                covariance_type = config['covariance'],
                                random_state = np.random.randint(0, 1e9),
                                ).fit(X)
    # print('means', skl_model.means_)
    labels_sk2 = skl_model2.predict(X)
    
    # Check accordance, taking the max in case labels are swapped
    return max(sum(labels_sk==labels_sk2), sum(labels_sk!=labels_sk2) ) / len(X) 



accord= []; sk_sk_accord = []; pycav_accord =[]
for i in range(10):
    accord.append(pycave_vs_sklearn_gmm(X, config))
    sk_sk_accord.append(sklearn_vs_sklearn_gmm(X, config))
    pycav_accord.append(pycave_vs_pycave_gmm(X, config))

print('Accordance over 10 experiments')
print("Sklearn vs Sklearn GMM Accordance :  % 2.2f %%" %(100*np.mean(sk_sk_accord)))
print("PyCave vs Pycave GMM Accordance :  % 2.2f %%" %(100*np.mean(accord)))
print("PyCave vs Sklearn GMM Accordance :  % 2.2f %%" %(100*np.mean(pycav_accord)))

```

```
Accordance over 10 experiments
Sklearn vs Sklearn GMM Accordance :   100.00 %
PyCave vs Pycave GMM Accordance :   76.47 %
PyCave vs Sklearn GMM Accordance :   86.23 %
```

Unless I'm doing something wrong, PyCave GMM accordance is less than 90%.

**Comment by morawi** [comment]
Update: 
Accordance over 100 experiments
Sklearn vs Sklearn GMM Accordance :   99.79 %
PyCave vs Pycave GMM Accordance :   77.06 %
PyCave vs Sklearn GMM Accordance :   79.50 %

**Comment by borchero** [comment]
I will have a look on the weekend :)

**Comment by morawi** [comment]
> I will have a look on the weekend :)
Update: All above are for K=2, for higher K values, PyCave did quite well. Here's one experiment using K=9

```
Accordance over 10 experiments
Sklearn vs Sklearn GMM Accordance :   78.12 %
PyCave vs Pycave GMM Accordance :   91.12 %
PyCave vs Sklearn GMM Accordance :   91.89 %

```

**Comment by borchero** [comment]
Ah I noticed that sklearn uses k-means initialization by default, so this might not be a fair comparison. Also, I'm not sure what kind of random initialization sklearn uses, so you might observe differences even if you do this.
